### ISSUE TITLE
Temporary restic files not cleaned up after crash or bad exit (CRITICAL)

### ISSUE THREAD

**Issue by yhafri** [restic/restic#873]
## Output of `restic version`

`Any`

## Expected behavior

Add option *--tmp_dir* to force restic to use a predefined directory.
This will bring to advantages:
1. being able to cleanup the temporary files in case it crashes
2. use temportay dir when we know we've enough disk space

## Actual behavior

Restic's temporary directory is system dependent:

```
Windows:
C:/Users/USER/AppData/Local/Temp/

Linux:
/tmp/...

MacOS:
$ ls /var/folders/vp/fw3j6w856bz5nwbj287rscq80000gn/T/restic-temp-pack-*
/var/folders/vp/fw3j6w856bz5nwbj287rscq80000gn/T/restic-temp-pack-802313766
/var/folders/vp/fw3j6w856bz5nwbj287rscq80000gn/T/restic-temp-pack-803271187
/var/folders/vp/fw3j6w856bz5nwbj287rscq80000gn/T/restic-temp-pack-823430765
/var/folders/vp/fw3j6w856bz5nwbj287rscq80000gn/T/restic-temp-pack-830939655
/var/folders/vp/fw3j6w856bz5nwbj287rscq80000gn/T/restic-temp-pack-846174674
/var/folders/vp/fw3j6w856bz5nwbj287rscq80000gn/T/restic-temp-pack-850278897
/var/folders/vp/fw3j6w856bz5nwbj287rscq80000gn/T/restic-temp-pack-851739399
/var/folders/vp/fw3j6w856bz5nwbj287rscq80000gn/T/restic-temp-pack-895818992
/var/folders/vp/fw3j6w856bz5nwbj287rscq80000gn/T/restic-temp-pack-913601601
/var/folders/vp/fw3j6w856bz5nwbj287rscq80000gn/T/restic-temp-pack-914202440
/var/folders/vp/fw3j6w856bz5nwbj287rscq80000gn/T/restic-temp-pack-937414976
/var/folders/vp/fw3j6w856bz5nwbj287rscq80000gn/T/restic-temp-pack-954527151
/var/folders/vp/fw3j6w856bz5nwbj287rscq80000gn/T/restic-temp-pack-955088802
/var/folders/vp/fw3j6w856bz5nwbj287rscq80000gn/T/restic-temp-pack-971129473
/var/folders/vp/fw3j6w856bz5nwbj287rscq80000gn/T/restic-temp-pack-987045776
[...]
```

## Steps to reproduce the behavior

Sometimes, restic crashes with errors like this one under MacOS, Linux, and Windows:

```
Windows:
[0:00] 1 directories, 1 files, 420.411 MiB
panic: client.PutObject: Put http://127.0.0.1:2008/USER/restic/data/e1dbb1bb89188ce12d38e8bb1dbac4e2b62554e265c1da1063473e454ed5a1c7: read tcp 127.0.0.1:57019->127.0.0.1:2008: wsarecv: An existing connection was forcibly closed by the remote host.
goroutine 362 [running]:
panic(0x854740, 0xc04a159940)
C:/Go/src/runtime/panic.go:500 +0x1af
restic/archiver.(*Archiver).saveChunk(0xc04200abe0, 0xc209ae1, 0x18db84, 0x1c1a8f7bc00000, 0xc04af3a000, 0x18db84, 0x1ee000, 0xc04429a210, 0xa99c40, 0xc042062140, ...)
C:/Users/USER/AppData/Local/Temp/restic-build-577404351/src/restic/archiver/archiver.go:163 +0x289
created by restic/archiver.(*Archiver).SaveFile
C:/Users/USER/AppData/Local/Temp/restic-build-577404351/src/restic/archiver/archiver.go:237 +0x392

MacOS:
[0:00] 1 directories, 3 files, 925.549 MiB
scanned 1 directories, 3 files in 0:00
panic: client.PutObject: Put http://xxxxxxx:2009/USER/restic/data/9e2bff1f6cd06015849b1aa65a0bc69d9e88dbed959f2720e3cb8141ee6d4c4d: read tcp 192.168.1.102:58650->XXX.XXX.XXX.XXX:2009: read: operation timed out
goroutine 280 [running]:
panic(0x480280, 0xc42000edc0)
/usr/local/Cellar/go/1.7.3/libexec/src/runtime/panic.go:500 +0x1a1
restic/archiver.(*Archiver).saveChunk(0xc420228500, 0x75fab0e, 0x1dd600, 0x5af040d800000, 0xc42b3bc000, 0x1dd600, 0x1ee000, 0xc4222e2210, 0x6edc40, 0xc4202b6008, ...)
/var/folders/vp/fw3j6w856bz5nwbj287rscq80000gn/T/restic-build-453067621/src/restic/archiver/archiver.go:163 +0x282
created by restic/archiver.(*Archiver).SaveFile
/var/folders/vp/fw3j6w856bz5nwbj287rscq80000gn/T/restic-build-453067621/src/restic/archiver/archiver.go:237 
```

when this happens, the temporary directories are never cleaned up.

**Comment by sanmai** [comment]
I've seen the same issue after running failed tests.

**Comment by fd0** [comment]
Thanks for reporting this issue. I'm not sure why you tag this as "CRITICAL", since (at least on Unix systems) the temp dir is regularly cleaned up. I suppose it's the same for Windows.

There are a few different things at play here:
 * First, you can set the temporary directory globally (for all Go programs) via the environment variable `TMPDIR`, restic respects that. Since the question came up a few times, I'd like to document this variable, but I'm not sure yet where it should be placed. Any ideas?
 * restic crashes with a panic because there's no retry logic for backend saves, that's tracked in #353 (I admit that the title for that issue is sub-optimal)
 * In addition, I'm currently rewriting the archiver code to better handle errors (instead of throwing a panic), so that we get the chance to e.g. clean up temp files.

So, in summary I think this issue is mainly about documenting the `TMPDIR` environment variable, the rest is either being worked on (see PR #842). I'm labeling this issue as `documentation`.

Anything to add?

**Comment by middelink** [comment]
Regarding files left behind in testing. I noticed that if your test panics, which mine did last night, the Go test framework itself will recover, but the cleanup will not be run. So to fix that, i guess we have to add panic recovery to withTestEnvironment?

**Comment by fd0** [comment]
Hm, I'm not sure about that. If the test panics, that's a pretty severe error... I'll look into it. Which test was it?

**Comment by middelink** [comment]
TestFindJSON(), but it is fixed now, you don't think I would submit a failing test? :P

btw, I like this way of setting up:

func testBla() {
   teardown := withEnvironment(t...)
   defer teardown(t)

   ... testcode ...
}
defer makes sure the teardown code gets run, even in light of a panic attack.

**Comment by fd0** [comment]
Oh, you're right! I think we should rework the tests so that `defer()` with a cleanup function is always used (instead of `withTestEnvironment()`, which I never really liked because it adds a level of indentation...).

**Comment by yhafri** [comment]
I've tagged the issue as CRITICAL because on our environment, restic crashes a lot and the temp data accumulated and never get deleted. As a consequence, we had issue with disk space.

But it's ok for the `TMPDIR` env. We'll play with it next week and let you know.

**Comment by fd0** [comment]
Ok, please let me know if you have an idea how to document this better.

**Comment by yhafri** [comment]
@fd0 why not simply provide us with an example? A simple use case.

**Comment by fd0** [comment]
The question is: Where should this example go, so that people like you find it? Where did you try to look for it?

**Comment by yhafri** [comment]
I always search things in the [user manual](https://github.com/restic/restic/blob/master/doc/Manual.md).

There's no better place IMHO.

**Comment by mholt** [comment]
> (at least on Unix systems) the temp dir is regularly cleaned up. I suppose it's the same for Windows.

Not a criticism, but I've always been told (and have adopted this belief) that it's the program's job to clean up its temporary files, not the OS's. For what my two cents are worth: I tend to agree, and especially in Go, since it's very easy to clean up with `defer`.

**Comment by fd0** [comment]
You're right and I agree, we'll work towards this.
